<a href="https://colab.research.google.com/github/EleanorNg/Data-Analysis-projects/blob/main/Project_Intro_to_Pandas_%5BFilters_%26_Groupby%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Immoscout](https://raw.githubusercontent.com/juliandnl/redi_ss20/master/image.png)

# Project - Intro to Pandas [Filter and Group]

--- 
## The description:
Hi all! You're already familiar with the Immobilienscout 24 dataset from the last project. In this project, we will explore the dataset further by sorting, filtering and grouping.

--- 
## The dataset:
The datasets contains information about rental appartments in Berlin. 
One row is for one appartment. The dataset contains 6 columns.  
- Region: is the Berlin district, where the flat is located
- Condition: is the condition of the flat
- Rooms: The number of rooms the flat has
- Rent: Monthly rent for the flat
- Year_Construction: The year in which the house was build
- Space: How many square meters does the flat has?

--- 
## The Task:

1. How many apartments are there in Mitte in a 'well-kept' condition?
2. How many apartments are there in Kreuzberg built before 1900?
3. What are the average spaces of the apartments having 4 rooms?
4. Which is the most expensive region on average and what is the maximum rent for this region?

Bonus:
5. What are the min, max and average number of rooms of the apartments in Köpenick?


In [ ]:
import pandas as pd
from scipy import stats

In [ ]:
rental=pd.read_csv("https://raw.githubusercontent.com/juliandnl/redi_ss20/master/berlin_rental.csv")

In [ ]:
rental.head()

Region       Condition  Rooms     Rent  Year_Construction   Space
0        Mitte  first_time_use    4.0  2659.00               2019  117.20
1    Kreuzberg  first_time_use    1.0  1200.00               2020   29.33
2     Köpenick       well_kept    2.0   979.00               1997   83.61
3  Wilmersdorf       well_kept    4.0  1830.22               1900  171.18
4    Kreuzberg  first_time_use    2.0  2272.00               2020   88.27

# 1.How many apartments are there in Mitte in a 'well-kept' condition?

In [ ]:
rental[(rental['Region']=="Mitte")&(rental['Condition']=="well_kept")]

Region  Condition  Rooms     Rent  Year_Construction   Space
15   Mitte  well_kept    2.0  1505.60               2018   75.28
272  Mitte  well_kept    3.0  1639.00               2014   96.00
392  Mitte  well_kept    1.0   924.44               2000   61.14
608  Mitte  well_kept    3.0  1800.00               2013   80.95
735  Mitte  well_kept    2.0  1350.00               2014   68.00
736  Mitte  well_kept    3.0  3500.00               2009  165.00

In [ ]:
rental[(rental['Region']=="Mitte")&(rental['Condition']=="well_kept")].shape[0]

6

There are 6 Apartments in Berlin-Mitte in in a 'well-kept'condition

# 2.How many apartments are there in Kreuzberg built before 1900?

In [ ]:
rental.query('Region=="Kreuzberg"&Year_Construction<1900')

Region                           Condition  Rooms    Rent  \
76   Kreuzberg                      first_time_use    5.0  4400.0   
401  Kreuzberg                         refurbished    4.0  2850.0   
697  Kreuzberg  first_time_use_after_refurbishment    2.0  1399.0   

     Year_Construction  Space  
76                1867  236.0  
401               1895  127.0  
697               1874   50.0

In [ ]:
rental.query('Region=="Kreuzberg"&Year_Construction<1900').shape[0]

3

There are 3 apartments in Kreuzberg built before 1900

# 3.What are the average spaces of the apartments by the number of rooms?

In [ ]:
x=rental['Space'].mean()
x

84.65626963350782

In [ ]:
y=rental['Rooms'].sum()
y

1948.0

In [ ]:
x*y

164910.41324607324

In [ ]:
rental.groupby('Rooms')[['Space']].mean()

Space
Rooms            
1.0     34.470748
1.5     41.361111
2.0     65.015725
2.5     90.463125
3.0     99.403167
3.5     98.037500
4.0    130.863333
4.5    106.100000
5.0    194.400526
6.0    237.000000
7.0    251.000000

The average spaces of the apartments by number of rooms is 164910.41324607324 

# 4.Which is the most expensive region on average and what is the maximum rent for this region?

In [ ]:
rental.head()

Region       Condition  Rooms     Rent  Year_Construction   Space
0        Mitte  first_time_use    4.0  2659.00               2019  117.20
1    Kreuzberg  first_time_use    1.0  1200.00               2020   29.33
2     Köpenick       well_kept    2.0   979.00               1997   83.61
3  Wilmersdorf       well_kept    4.0  1830.22               1900  171.18
4    Kreuzberg  first_time_use    2.0  2272.00               2020   88.27

In [ ]:
rental.groupby('Region')[['Rent']].mean().sort_values(by="Rent",ascending=False)

Rent
Region                      
Mitte            2408.725033
Kreuzberg        2049.138085
Prenzlauer       1996.962264
Tiergarten       1914.824795
Charlottenburg   1753.452532
Wilmersdorf      1748.646226
Neukölln         1188.114915
Köpenick         1125.475818
Friedrichsfelde  1043.871731
Wedding           847.138571

In [ ]:
rental.groupby('Region')[['Rent']].mean().sort_values(by="Rent",ascending=False).iloc[0]

Rent    2408.725033
Name: Mitte, dtype: float64

In [ ]:
rental.groupby('Region')[['Rent']].agg(['max'])

Rent
                      max
Region                   
Charlottenburg    9774.00
Friedrichsfelde   2138.68
Kreuzberg         6450.00
Köpenick          2159.20
Mitte            14207.00
Neukölln          2295.00
Prenzlauer        5930.00
Tiergarten        4945.00
Wedding           1949.26
Wilmersdorf       5876.35

The most expensive Region on average in Berlin is Mitte and the maximum Rent is 14207.00 Euros 

# Bonus Question:
5. What are the min, max and average number of rooms of the apartments in Köpenick?


In [ ]:
rental.groupby('Region')[['Rooms']].agg(['min','max','mean'])

Rooms               
                  min  max      mean
Region                              
Charlottenburg    1.0  5.0  2.525316
Friedrichsfelde   1.0  4.5  2.500000
Kreuzberg         1.0  5.0  2.489362
Köpenick          1.0  4.0  2.281818
Mitte             1.0  5.0  2.754902
Neukölln          1.0  5.0  2.279661
Prenzlauer        1.0  6.0  2.660377
Tiergarten        1.0  5.0  2.736842
Wedding           1.0  4.0  1.726190
Wilmersdorf       1.0  7.0  2.613208

In [ ]:
rental[rental['Region']=='Köpenick']['Rooms'].agg(['min','max','mean']).round(2)

min     1.00
max     4.00
mean    2.28
Name: Rooms, dtype: float64

The min number of rooms of the apartments in Köpenick is 1,
The max number of rooms of the apartments in Köpenick is 4 and 
The average number of rooms apartments in Köpenick is 2.28.

In [ ]:
#Transformations:	
#Is there a difference in average room space (Space divided by Rooms) for apartments built after 2000 versus apartments built before 2000?
rental["Average_Room_space"] = (rental.Space / rental.Rooms).round(1)
rental.head()


Region       Condition  Rooms     Rent  Year_Construction   Space  \
0        Mitte  first_time_use    4.0  2659.00               2019  117.20   
1    Kreuzberg  first_time_use    1.0  1200.00               2020   29.33   
2     Köpenick       well_kept    2.0   979.00               1997   83.61   
3  Wilmersdorf       well_kept    4.0  1830.22               1900  171.18   
4    Kreuzberg  first_time_use    2.0  2272.00               2020   88.27   

   Average_Room_space  
0                29.3  
1                29.3  
2                41.8  
3                42.8  
4                44.1

In [ ]:
rental_After_2000 = rental[rental.Year_Construction > 2000]
rental_After_2000.head()

Region       Condition  Rooms     Rent  Year_Construction   Space  \
0       Mitte  first_time_use    4.0  2659.00               2019  117.20   
1   Kreuzberg  first_time_use    1.0  1200.00               2020   29.33   
4   Kreuzberg  first_time_use    2.0  2272.00               2020   88.27   
6       Mitte  mint_condition    2.0  1509.45               2015   61.61   
9  Tiergarten  first_time_use    3.0  1789.00               2020   77.66   

   Average_Room_space  
0                29.3  
1                29.3  
4                44.1  
6                30.8  
9                25.9

In [ ]:
rental_Before_2000 = rental[rental.Year_Construction < 2000]
rental_Before_2000.head()

Region  Condition  Rooms     Rent  Year_Construction   Space  \
2         Köpenick  well_kept    2.0   979.00               1997   83.61   
3      Wilmersdorf  well_kept    4.0  1830.22               1900  171.18   
5         Köpenick  well_kept    2.0   840.00               1997   73.51   
7   Charlottenburg  well_kept    2.0   730.73               1900   72.61   
8  Friedrichsfelde  well_kept    2.0   561.93               1971   62.40   

   Average_Room_space  
2                41.8  
3                42.8  
5                36.8  
7                36.3  
8                31.2

In [ ]:
Sample_A = rental_After_2000.Average_Room_space
Sample_A.head()

0    29.3
1    29.3
4    44.1
6    30.8
9    25.9
Name: Average_Room_space, dtype: float64

In [ ]:
Sample_B = rental_Before_2000.Average_Room_space
Sample_B.head()

2    41.8
3    42.8
5    36.8
7    36.3
8    31.2
Name: Average_Room_space, dtype: float64

In [ ]:
print("Null_Hypothesis = (Sample_A = Sample_B)")

Null_Hypothesis = (Sample_A = Sample_B)


In [ ]:
t_results = stats.ttest_ind(Sample_A, Sample_B)
t_results

Ttest_indResult(statistic=-1.7131027560199323, pvalue=0.08710147155678297)

In [ ]:
alpha = 0.05
print(alpha)

0.05


In [ ]:
if (t_results[1] < alpha):
  print("Sample_A and Sample_B are different")

In [ ]:
#How many apartments were posted each month?
Yearly_Apartment_Posted = rental.groupby('Year_Construction')[["Rooms"]].count()
Yearly_Apartment_Posted.tail()

Rooms
Year_Construction       
2016                  28
2017                  45
2018                  58
2019                 181
2020                  58

In [ ]:
Monthly_Apartment_Posted = (Yearly_Apartment_Posted/12).round(1)
Monthly_Apartment_Posted.tail()

Rooms
Year_Construction       
2016                 2.3
2017                 3.8
2018                 4.8
2019                15.1
2020                 4.8

In [ ]:
#Bonus:
#Can you think about other KPIs for Immobilienscout? (It doesn't need to use only the data from this dataset)


--- 
# Review

This section is only for the reviewing team!

Guideline how to review:
https://docs.google.com/presentation/d/1YORFwlfVQo9ogj7jR9t6_pqxmGIlpBSGubbp1UdtcBQ/edit?usp=sharing


In [ ]:
#@title Review Criteria:
%%html
<h3><input type="checkbox"> 1. Are all questions answered? <br></h3> 
<h3><input type="checkbox"> 2. Does all code run through? <br></h3> 
<h3><input type="checkbox"> 3. Are the conclusions understandable?  <br></h3> 
<h3><input type="checkbox"> 4. Are the bonus question answered?  <br></h3> 

In [ ]:
input("Do you have any feedback?")